nengo/utils/tests/test_magic.py

```diff
@@ -228,3 +232,149 @@ def __init__(self, a, b):
         "            self.a = a\n"
         "            self.b = b\n"
     )
+
+
+def test_class_decorator():
+    """Test that `decorator` works on a class method"""
+
+    class TestA:
+        @decorator
+        @classmethod
+        def test_decorator(cls, wrapped, instance, args, kwargs):
+            global state
+            state = "run"
+            assert instance is not None
+            assert type(instance).__name__ == "Test"
+            return wrapped(*args, **kwargs)
+
+    class Test:
+        @TestA.test_decorator
+        def f(self, a, b):
+            """Return 1."""
+            return 1
+
+    inst = Test()
+    _test_decorated(inst.f)
+
+    # Make sure introspection works
+    assert inspect.getfullargspec(inst.f).args == ["self", "a", "b"]
+    assert inspect.getsource(inst.f) == (
+        "        @TestA.test_decorator\n"
+        "        def f(self, a, b):\n"
+        '            """Return 1."""\n'
+        "            return 1\n"
+    )
+
+
+def test_instance_decorator():
+    """Test that `decorator` works on a class instance method"""
+
+    class TestA:
+        @decorator
+        def test_decorator(self, wrapped, instance, args, kwargs):
+            global state
+            state = "run"
+            assert instance is not None
+            assert type(instance).__name__ == "Test"
+            return wrapped(*args, **kwargs)
+
+    test_a = TestA()
+
+    class Test:
+        @test_a.test_decorator
+        def f(self, a, b):
+            """Return 1."""
+            return 1
+
+    inst = Test()
+    _test_decorated(inst.f)
+
+    # Make sure introspection works
+    assert inspect.getfullargspec(inst.f).args == ["self", "a", "b"]
+    assert inspect.getsource(inst.f) == (
+        "        @test_a.test_decorator\n"
+        "        def f(self, a, b):\n"
+        '            """Return 1."""\n'
+        "            return 1\n"
+    )
+
+
+def test_objectproxy():
+    """tests functions of ObjectProxy"""
+
+    class Test:
+        """my docstring"""
+
+        __annotations__ = "testannotations"
+        __name__ = "testname"
+
+        testvar = "testval"
+
+    obj = Test()
+    proxy = ObjectProxy(obj)
+
+    # --- test ObjectProxyMethods
+    assert proxy.__dict__ is obj.__dict__
+    assert proxy.__doc__ is obj.__doc__
+    assert proxy.__module__ is obj.__module__
+
+    # --- test ObjectProxy
+    assert proxy.__annotations__ is obj.__annotations__
+    assert proxy.__name__ is obj.__name__
+    assert proxy.__class__ is obj.__class__
+
+    assert set(dir(proxy)) == set(dir(obj))
+    assert hash(proxy) == hash(obj)
+    assert str(proxy) == str(obj)
+
+    proxy.newattr = "newval"  # pylint: disable=assigning-non-slot
+    assert obj.newattr == "newval"
+    assert proxy.newattr == "newval"
+
+    assert repr(proxy).startswith("<ObjectProxy at ")
+
+
+def test_boundfunctionwrapper():
```

Comment:
It would be nice to have a more functional test of BoundFunctionWrapper, but I can't figure out how to use a lot of its functionality. I can't trigger the `__get__` function at all, and I also can't seem to get `self.instance is None`. Any ideas @tbekolay?

Reply:
I really don't think it's that important; everything in the `magic.py` file is from [wrapt](https://pypi.org/project/wrapt/), which has its own test suite. We could vendorize it.